### ISSUE TITLE
assumption of Halide IR on a scalar variable use

### ISSUE THREAD

**Issue by ronghongbo** [halide/Halide#3000]
Hello,

Is it possible for Halide IR to have the following case?
```
   for i =.... {
         for j = ...  {
              ...= x+1
         }
         x = ..
  }
```

We suspect that this would never happen. That is, can we assert that in a valid Halide IR, a use of a scalar variable depends only on definitions LEXICALLY before it?

Thanks,
Hongbo

**Comment by abadams** [comment]
Yes, use of a scalar variable depends only on the definition immediately
enclosing (and precending) it in the IR. Variables introduced by let nodes
can't have their value changed - they're immutable. However, they *can* be
shadowed by variables with the same name in a nested scope.

let x = 5
let y = x + 1
for i = ... {
  let x = 8
  let y = x + 1
  // here x is 8 and y is 9
}
// here x is 5 and y is 6

We generally try to avoid doing this - in fact the simplifier refuses to
accept IR like this - but it does happen during the early stages of
lowering, so it counts as valid IR.

- Andrew

**Comment by abadams** [comment]
> [!NOTE]
> **Automated triage:** this issue was flagged by Claude as likely safe to close, for the
> reason below. The reasoning was checked against the current source, and a second pass
> tried and failed to refute it, but no human verified this specific text.
> **If the reason is wrong, please reopen the issue and say so** — some fraction of these
> are expected to be wrong.

**Reason (CLOSE_STALE):** This was a question about IR semantics (can a scalar use refer to a definition lexically after it), not a bug report. abadams answered it definitively in the only comment: Let-bound variables are immutable and a use refers to the enclosing/preceding definition, with shadowing possible during early lowering.

**Evidence:** Issue thread: single answer from abadams 2018-05-30; no code change was ever requested. Semantics unchanged — Let/LetStmt in src/IR.h still bind immutable values.

Closing: this was a question about IR semantics, answered in-thread — a use of a scalar variable always refers to the immediately enclosing, lexically preceding Let/LetStmt (values are immutable; shadowing by a nested scope is possible in early lowering). No code change tracked here.
